### PR TITLE
Enable cognito login for admin app users

### PR DESCRIPTION
**Pull request by danjoneslf**
Allow/deny routes based on user role instead of ADMIN env var.
Deny admin users access to files/downloads routes.
Change welcome template to show appropriate content for admin users.

I've also done a quick drive by to remove the zipped lambda file and ignore which it looks like I accidentally added when we moved repos.

### CODE REVIEW

File: behave/features/steps/user_routes.py [hunk 1]
```diff
@@ -83,11 +116,18 @@ def user_path_step(context, path):
...
-def user_redirect_step(context):
+def user_redirect_home_step(context):
     url = os.environ["E2E_STAGING_ROOT_URL"]
     assert context.browser.current_url == url
 
 
+@then('you get redirected to route: "{route}"')
+def user_redirect_to_route_step(context, route):
+    url = re.sub("/$", route, os.environ["E2E_STAGING_ROOT_URL"])
+    print(url)
```
Comment (oscarwyatt) [inline behave/features/steps/user_routes.py:127]: I suspect we can remove the print

File: flask_helpers.py [hunk 2]
```diff
@@ -1,33 +1,108 @@
...
 
 def is_admin_interface():
     return os.getenv("ADMIN", "false") == "true"
 
 
+def has_admin_role():
+    is_admin_role = False
+    group = session.get("group", None)
+    if group:
+        is_admin_role = "admin-" in group["value"]
```
Comment (oscarwyatt) [inline flask_helpers.py:18]: I suspect you'll know more than I about this kind of thing so feel free to disagree. While I'll concede it's unlikely, it is theoretically possible that in the future someone could add a new group containing "admin-" (say "local-authority-admin-barnet") that isn't meant to have admin control over the entire app (in this example, just a subset of users) and wouldn't necessarily check through the code to see if it has side effects - causing unintended security problems. I suggest this be changed to something like this which is more explicit:
```
is_admin_role = group["value"] in ["admin-view",  "admin-power", "admin-full"]
```

File: flask_helpers.py [hunk 3]
```diff
@@ -1,33 +1,108 @@
...
+        if has_admin_role():
+            # raise Exception("User not authorised to access this route")
+            session["error_message"] = "User not authorised to access this route"
+            return redirect("/403")
+        return flask_route(*args, **kwargs)
+
+    return decorated_function
+
+
+def requires_group_matching(role_like):
```
Comment (oscarwyatt) [inline flask_helpers.py:69]: For the same reason as above (and with the same caveat as you probably knowing more than I) feel uneasy about using regexes for these matches, there are only six user groups so an explicit list is (currently) always going to be short and will avoid any unintended consequences

File: flask_helpers.py [hunk 4]
```diff
@@ -1,33 +1,108 @@
...
+
+    return decorate_route
+
+
+def requires_group_in_list(valid_roles):
+    def decorate_route(flask_route):
+        @wraps(flask_route)
+        def decorated_function(*args, **kwargs):
+            if current_group_name() not in valid_roles:
+                # raise Exception("User not authorised to access this route")
```
Comment (oscarwyatt) [inline flask_helpers.py:89]: if this line isn't needed/doesn't work would be good to remove it

File: flask_helpers.py [hunk 5]
```diff
@@ -1,33 +1,108 @@
...
+
+    return decorated_function
+
+
+def requires_group_matching(role_like):
+    def decorate_route(flask_route):
+        @wraps(flask_route)
+        def decorated_function(*args, **kwargs):
+            if not bool(re.search(role_like, current_group_name())):
+                # raise Exception("User not authorised to access this route")
```
Comment (oscarwyatt) [inline flask_helpers.py:74]: if this line isn't needed/doesn't work would be good to remove it

File: flask_helpers.py [hunk 6]
```diff
@@ -1,33 +1,108 @@
...
         return flask_route(*args, **kwargs)
 
     return decorated_function
 
 
+def end_user_interface(flask_route):
+    @wraps(flask_route)
+    def decorated_function(*args, **kwargs):
+        if has_admin_role():
+            # raise Exception("User not authorised to access this route")
```
Comment (oscarwyatt) [inline flask_helpers.py:61]: if this line isn't needed/doesn't work would be good to remove it

File: flask_helpers.py [hunk 7]
```diff
@@ -37,7 +112,9 @@ def upload_rights_required(flask_route):
     @wraps(flask_route)
     def decorated_function(*args, **kwargs):
         if not has_upload_rights():
-            return redirect("/")
+            # return redirect("/")
```
Comment (oscarwyatt) [inline flask_helpers.py:115]: if this line isn't needed/doesn't work would be good to remove it

File: flask_helpers.py [hunk 8]
```diff
@@ -1,33 +1,108 @@
...
+
+    return decorate_route
+
+
 def login_required(flask_route):
     @wraps(flask_route)
     def decorated_function(*args, **kwargs):
         if "details" not in session:
-            return redirect("/")
+            # return redirect("/")
```
Comment (oscarwyatt) [inline flask_helpers.py:103]: if this line isn't needed/doesn't work would be good to remove it

File: behave/features/steps/user_routes.py (hunk 1 above)
Comment (danjoneslf) [reply behave/features/steps/user_routes.py:127]: 👍

File: flask_helpers.py (hunk 2 above)
Comment (danjoneslf) [reply flask_helpers.py:18]: Good catch. I'll change.

File: flask_helpers.py (hunk 3 above)
Comment (danjoneslf) [reply flask_helpers.py:69]: Agreed.

File: flask_helpers.py (hunk 4 above)
Comment (danjoneslf) [reply flask_helpers.py:89]: Agreed - I was leaving that in as a comparison against what was happening before. I'll remove.

**Pull request merged**